### DISCUSSION TITLE
Is there a way to hook/enumerate that a Body has ended?

### DISCUSSION THREAD

**Discussion by AshleySchaeffer** [hyperium/hyper#2785] [Q&A] [answered]
I'd like to do further computations upon the completion of sending/receiving a hyper Body.

I'm not 100% sure what the best approach to this is after playing with available methods and some debugging.

Anyone got any ideas?

**Comment by seanmonstar** [comment, marked as answer]
When _receiving_ a body, if you're polling the data of it, you can just trigger a computation once it returns `None`. Sending would require you to instrument an `impl HttpBody`. The [`tower_http::trace` body wrapper](https://github.com/tower-rs/tower-http/blob/master/tower-http/src/trace/body.rs) might help show you. It wraps any other body type with potential callbacks as chunks are emitted and as the body ends.

**Reply by AshleySchaeffer** [reply]
Thanks for replying so quickly. 

I have a setup that's polling a receiving Body and it never seems to return None. Would you mind providing a crude example for my sanity? I've been going at this all day.

**Reply by AshleySchaeffer** [reply]
I'm also a bit wary of implementing my own variant of Body as it looks like there's some client related stuff in the hyper one that might be important (delayed EOF). 

How safe is it to implement my own Body?

**Reply by seanmonstar** [reply]
At the bottom of the [client guide](https://hyper.rs/guides/client/basic/), you can see how to iterate the data of a body, and when that `while let Some(..)` loop finishes, the body has finished.

It's safe to implement the `hyper::body::HttpBody` trait. You don't need to worry about the internal stuff of the `hyper::Body` struct, it's doing special stuff.

**Reply by AshleySchaeffer** [reply]
Thank you. I think I'll try a wrapper and see if I can use it for both sending receiving.

To be clear, and I'll stop nagging after this, will I definitely not see any differences between `hyper::Body` and my own? Sorry to question it again, I'm mainly concerned by this stuff (see comment). I am using the hyper client: https://github.com/hyperium/hyper/blob/740654e55d2bb2f50709f20fb4054a5504d8c2fb/src/body/body.rs#L67-L77
